Bed Fix

SwivelDiscourse opened the pull request:
Fixed Sleeping in same Bed as Citizen Bug

Closes #2656 

# Changes proposed in this pull request:
-Updates EventHandler.java to check OnActivated Event for Player to see if player is trying to sleep in bed with a citizen in it, and cancels the event if true.

Review please

Commit:
Bed Fix

Fixed Sleeping in same Bed as Citizen Bug

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ldtteam/minecolonies?pullRequest=2936) <br/>All committers have signed the CLA.

LDTTeam-Buildsystem reviewed (dismissed):
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  |   0   |  0  |  0  |  0   |
| Errors |   0   |  0  |  0  |  0   |
##### More Information:

LDTTeam-Buildsystem reviewed (dismissed):
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  |   0   |  0  |  0  |  0   |
| Errors |   0   |  0  |  0  |  0   |
##### More Information:

Raycoms commented on src/main/java/com/minecolonies/coremod/event/EventHandler.java:
```diff
@@ -309,6 +314,22 @@ else if ("pmardle".equalsIgnoreCase(event.getEntityPlayer().getName())
+                    bd = bd.offset(world.getBlockState(event.getPos()).getValue(BlockBed.FACING));
+                }
+                //Searches through the nearest Colonie's Citizen and sees if the bed belongs to a Citizen, and if the Citizen is asleep
+                final ArrayList<CitizenData> cits = new ArrayList<>(ColonyManager.getClosestColony(world, event.getPos()).getCitizenManager().getCitizens());
```
List<CitizenData> cits = new ArrayList<>();
Also, better save the colony and check for null first, if colony is null no need for all of this.

Raycoms commented on src/main/java/com/minecolonies/coremod/event/EventHandler.java:
```diff
@@ -309,6 +314,22 @@ else if ("pmardle".equalsIgnoreCase(event.getEntityPlayer().getName())
+                }
+                //Searches through the nearest Colonie's Citizen and sees if the bed belongs to a Citizen, and if the Citizen is asleep
+                final ArrayList<CitizenData> cits = new ArrayList<>(ColonyManager.getClosestColony(world, event.getPos()).getCitizenManager().getCitizens());
+                for(int i = 0; i < cits.size(); i++){
```
for each?

Raycoms reviewed (dismissed):
Nice changes, a few small additions:

Can you please update the formatting to fit the general style?
Especially the bracket style, there is a formatter xml file in the repo you can download and import with intellij

ghost commented:
Will Do. Give me a day or so to fix things over.

Nightenom commented on src/main/java/com/minecolonies/coremod/event/EventHandler.java:
```diff
@@ -309,6 +314,22 @@ else if ("pmardle".equalsIgnoreCase(event.getEntityPlayer().getName())
+             //Checks to see if player tries to sleep in a bed belonging to a Citizen, ancels the event, and Notifies Player that bed is occuppied
+            if(world.getBlockState(event.getPos()).getBlock().isBed(world.getBlockState(event.getPos()), world, event.getPos(), player))
+            {
+                BlockPos bd = event.getPos();
```
Move it before the `if` if possible (and use it whenever possible), also rename to something more understandable

Nightenom replied:
edit: move it on 289 where are finals for this event, on 322 create a new variable with an appropriate name

LDTTeam-Buildsystem reviewed (dismissed):
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  |   0   |  0  |  0  |  0   |
| Errors |   0   |  0  |  0  |  0   |
##### More Information:

LDTTeam-Buildsystem reviewed (dismissed):
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  |   0   |  0  |  0  |  0   |
| Errors |   0   |  0  |  0  |  0   |
##### More Information:

Commit:
Updated Bed Fix

Changed Formatting, Matched Code Styles, fixed overall code format, layout and design choices to match project.
Added Null Check for event to check Colony

LDTTeam-Buildsystem reviewed (approved):
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  |   0   |  0  |  0  |  0   |
| Errors |   0   |  0  |  0  |  0   |
##### More Information:

LDTTeam-Buildsystem reviewed (dismissed):
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  |   0   |  0  |  0  |  0   |
| Errors |   0   |  0  |  0  |  0   |
##### More Information:

Commit:
Bed Fix 3.0

Fixed Formatting so Codacy-senpai might actually approve of me this time.

Raycoms commented on src/main/java/com/minecolonies/coremod/event/EventHandler.java:
```diff
@@ -286,6 +293,9 @@ public void onPlayerInteract(@NotNull final PlayerInteractEvent.RightClickBlock
         final EntityPlayer player = event.getEntityPlayer();
         final World world = event.getWorld();
+        BlockPos bedBlockPos = event.getPos();
+        final List<CitizenData> citizenList;
```
I would get the event pos here yes, but I wouldn't put the world and citizen list here. Else we query the colony for each right click of a player, this is terrible performance wise

SwivelDiscourse replied:
The World object was already implemented in the code that way at the start, as it's used by several other events when the Player Right-Clicks. I'll move the Citizenlist object thought.

Raycoms replied:
sorry, I meant colony not world xD

Raycoms commented on src/main/java/com/minecolonies/coremod/event/EventHandler.java:
```diff
@@ -315,6 +325,27 @@ else if ("pmardle".equalsIgnoreCase(event.getEntityPlayer().getName())
             }
 
+                //Checks to see if player tries to sleep in a bed belonging to a Citizen, ancels the event, and Notifies Player that bed is occuppied
+                if (colony != null && world.getBlockState(event.getPos()).getBlock().isBed(world.getBlockState(event.getPos()), world, event.getPos(), player))
```
So first check for  bed and then check in the if for colony.

Raycoms commented on src/main/java/com/minecolonies/coremod/event/EventHandler.java:
```diff
@@ -315,6 +325,27 @@ else if ("pmardle".equalsIgnoreCase(event.getEntityPlayer().getName())
+                   citizenList = ColonyManager.getClosestColony(world, event.getPos()).getCitizenManager().getCitizens();
+                    if (world.getBlockState(event.getPos()).getBlock().isBedFoot(world, event.getPos()))
+                    {
+                     bedBlockPos = bedBlockPos.offset(world.getBlockState(event.getPos()).getValue(BlockBed.FACING));
```
this need indent

Commit:
Bed Fix 4.0

Ray likes his variables where they won't cause performance issues... for some reason.

Commit:
Bed Fix 5.0

sudo this is the final commit

Commit:
Merge branch 'version/1.12' into version/1.12

ghost commented:
Looks like it's all here. Codacy Likes it, format is good, Just waiting on approval.

Raycoms commented on src/main/java/com/minecolonies/coremod/event/EventHandler.java:
```diff
@@ -314,6 +322,30 @@ else if ("pmardle".equalsIgnoreCase(event.getEntityPlayer().getName())
+                //Checks to see if player tries to sleep in a bed belonging to a Citizen, ancels the event, and Notifies Player that bed is occuppied
+                if (colony != null)
+                {
+                    final List<CitizenData> citizenList = ColonyManager.getClosestColony(world, event.getPos()).getCitizenManager().getCitizens();
```
why don't you get a real colony above already, there is a ColonyManager.getColonyByPosFromWorld() then you don't have to get the colony twice.

SwivelDiscourse replied:
I don't see that function anywhere. I have GetClosestColony, and GetClosestIColony

Raycoms committed:
Merge branch 'version/1.12' into version/1.12

Commit:
Bed Fix 6.0

Updated Project, able to use cool new functions that make everything easier. Check for actual Colony now.

Raycoms commented:
:shipit:

[![Approved with PullApprove](https://img.shields.io/badge/maintainers-approved-brightgreen.svg)](https://pullapprove.com/Minecolonies/minecolonies/pull-request/2936/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)
